src/main/java/slimeknights/tconstruct/library/tools/ToolCore.java

```diff
@@ -142,290 +199,576 @@ public int getHarvestLevel(ItemStack stack, ToolType toolClass, @Nullable Player
     }
+
     return super.getDestroySpeed(stack, state);
   }
 
   @Override
   public boolean canHarvestBlock(ItemStack stack, BlockState state) {
-    return isEffective(state) && !ToolData.isBroken(stack);
+    return this.isEffective(state) && !ToolData.isBroken(stack);
+  }
+
+  /* Repairing */
+
+  @Override
+  public boolean needsRepair(ItemStack repairable) {
+    if (repairable.getDamage() == 0 && !ToolData.isBroken(repairable)) {
+      // undamaged and not broken - no need to repair
+      return false;
+    }
+
+    ToolData toolData = ToolData.readFromNBT(repairable.getTag());
+
+    List<IMaterial> materials = toolData.getMaterials();
+    return !materials.isEmpty();
+  }
+
+  @Nonnull
+  @Override
+  public ItemStack repair(ItemStack repairable, NonNullList<ItemStack> repairItems) {
+    //todo decide how to handle this
+    return repairable;
+  }
+
+  /* Attacking */
+
+  /**
+   * Actually deal damage to the entity we hit. Can be overridden for special behaviour
+   *
+   * @return True if the entity was hit. Usually the return value of {@link Entity#attackEntityFrom(DamageSource, float)}
+   */
+  public boolean dealDamage(ItemStack stack, LivingEntity player, Entity entity, float damage) {
+    if (player instanceof PlayerEntity) {
+      return entity.attackEntityFrom(DamageSource.causePlayerDamage((PlayerEntity) player), damage);
+    }
+
+    return entity.attackEntityFrom(DamageSource.causeMobDamage(player), damage);
+  }
+
+  /**
+   * Checks if the player is ready to special attack
+   *
+   * @param player the player to check
+   * @return whether or not if the player is ready for special attack
+   */
+  protected boolean readyForSpecialAttack(LivingEntity player) {
+    return player instanceof PlayerEntity && ((PlayerEntity) player).getCooledAttackStrength(0.5f) > 0.9f;
+  }
+
+  @Override
+  public boolean onLeftClickEntity(ItemStack stack, PlayerEntity player, Entity entity) {
+    return ToolAttackUtil.attackEntity(stack, this, player, entity);
+  }
+
+  @Override
+  public boolean hitEntity(ItemStack stack, LivingEntity target, LivingEntity attacker) {
+    float speed = ToolData.from(stack).getStats().attackSpeedMultiplier;
+
+    if (!stack.isEmpty() && stack.getItem() instanceof ToolCore) {
+      speed *= ((ToolCore) stack.getItem()).getToolDefinition().getBaseStatDefinition().getAttackSpeed();
+    }
+
+    int time = Math.round(20f / speed);
+    if (time < target.hurtResistantTime / 2) {
+      target.hurtResistantTime = (target.hurtResistantTime + time) / 2;
+      target.hurtTime = (target.hurtTime + time) / 2;
+    }
+
+    return super.hitEntity(stack, target, attacker);
+  }
+
+  @Override
+  public Multimap<Attribute, AttributeModifier> getAttributeModifiers(EquipmentSlotType slot, ItemStack stack) {
+    ImmutableMultimap.Builder<Attribute, AttributeModifier> builder = ImmutableMultimap.builder();
+
+    float speed = ToolData.from(stack).getStats().attackSpeedMultiplier;
```

Comment:
Create the `ToolData` once, saves the amount of NBT we have to parse. Then call methods from it later